astro_plasma/core/ionization.py

```diff
@@ -189,16 +190,31 @@ def interpolate_ion_frac(
             The value is in log10.
 
         """
+        if not (type(element) == AtmElement or type(element) == int):
+            elm_match = re.match(r"^([A-Z][a-z]?)([IVX]+)$", str(element).upper())
+            if elm_match:
+                _element, _ion_symbol = elm_match.groups()
+                ion = roman_to_int(_ion_symbol)
+            else:
+                raise ValueError(f"Invalid/Unsupported value of element: {element}.")
+            elm_atm_no = AtmElement.parse(str(_element)).to_atm_no()
+        elif type(element) == int:
+            elm_atm_no = element
+        elif type(element) == AtmElement:
+            elm_atm_no = element.to_atm_no()
+
         # element = 1: H, 2: He, 3: Li, ... 30: Zn
         # ion = 1 : neutral, 2: +, 3: ++ .... (element+1): (++++... element times)
-        if ion < 0 or ion > element + 1:
-            raise ValueError(f"Problem! Invalid ion {ion} for element {element}.")
-        if element < 0 or element > 30:
-            raise ValueError(f"Problem! Invalid ion {ion} for element {element}.")
+        if ion < 0 or ion > elm_atm_no + 1:
+            print("Problem! Invalid ion %d for element %d." % (ion, elm_atm_no))
+            return None
```

Comment:
Why would you not raise an exception?

Also if `None is to be returned, shouldn't the return type-hinting be changed to `Optional[float]`

Reply:
My bad it should raise exception

Reply:
> > @dutta-alankar Could you explain this change to me? Why is this required? [5d30891](https://github.com/dutta-alankar/AstroPlasma/commit/5d3089170a88d8cd098b907965aa5c327d4c05f6)
> 
> `mypy` type-hinting failure was happening without this refactoring.

The trick is to use. `not isinstance(element, AtmElement)` this will strip out `AtmElement` type from element in the if block scope.

And later ensure element is always `AtmElement`

Fixed in 88a712207c0a6ecb933d03079d3edb3790e68f86

I checked the following test cases are working

```python
i.interpolate_ion_frac(element=AtmElement.Aluminum, ion=10)
i.interpolate_ion_frac(element="C", ion=10)
i.interpolate_ion_frac(element=1, ion=10)
i.interpolate_ion_frac(element=10, ion=10)
i.interpolate_ion_frac(element="ZnXX")
```